netebpfext/net_ebpf_ext_xdp.c

```diff
@@ -45,28 +40,219 @@ DEFINE_GUID(EBPF_ATTACH_TYPE_XDP, 0x85e0d8ef, 0x579e, 0x4931, 0xb0, 0x72, 0x8e,
+    ASSERT(nbl != NULL);
+    net_buffer = NET_BUFFER_LIST_FIRST_NB(nbl);
+
+    if (delta == 0)
+        // Nothing to do.
+        goto Exit;
+    if (delta < 0) {
+        uint32_t absolute_delta = -delta;
+        ndis_status = NdisRetreatNetBufferDataStart(net_buffer, absolute_delta, 0, NULL);
+        if (ndis_status != NDIS_STATUS_SUCCESS) {
+            return_value = -1;
+            goto Exit;
+        }
+        packet_buffer = NdisGetDataBuffer(net_buffer, net_buffer->DataLength, NULL, 1, 0);
+        if (packet_buffer != NULL)
+            net_xdp_ctx->data = packet_buffer;
+        else {
+            // Data in net_buffer not contiguous.
+            // Restore net_buffer.
+            NdisAdvanceNetBufferDataStart(net_buffer, absolute_delta, TRUE, NULL);
+            // Allocate a cloned NBL with contiguous data.
+            _net_ebpf_ext_allocate_cloned_nbl(net_xdp_ctx, absolute_delta);
+        }
+    } else {
+        // delta > 0.
+        NdisAdvanceNetBufferDataStart(net_buffer, delta, FALSE, NULL);
+        packet_buffer = NdisGetDataBuffer(net_buffer, net_buffer->DataLength, NULL, 1, 0);
+        ASSERT(packet_buffer != NULL);
+        net_xdp_ctx->data = packet_buffer;
+    }
+
+Exit:
+    return return_value;
+}
+
+void
+_net_ebpf_ext_free_nbl(_Inout_ NET_BUFFER_LIST* nbl)
+{
+    NET_BUFFER* net_buffer = NET_BUFFER_LIST_FIRST_NB(nbl);
+    MDL* mdl_chain = NET_BUFFER_FIRST_MDL(net_buffer);
+    uint8_t* buffer = MmGetSystemAddressForMdlSafe(mdl_chain, NormalPagePriority);
+    if (buffer != NULL)
+        ExFreePool(buffer);
+    IoFreeMdl(mdl_chain);
+    FwpsFreeNetBufferList0(nbl);
+}
+
+static void
+_net_ebpf_ext_l2_receive_inject_complete(_In_ const void* context, _Inout_ NET_BUFFER_LIST* nbl, BOOLEAN dispatch_level)
 {
     UNREFERENCED_PARAMETER(context);
-    FwpsFreeCloneNetBufferList(packet_clone, dispatch_level);
+    UNREFERENCED_PARAMETER(dispatch_level);
+    // Free clone allocated using _net_ebpf_ext_allocate_cloned_nbl.
+    _net_ebpf_ext_free_nbl(nbl);
+}
+
+static FWP_ACTION_TYPE
+_net_ebpf_ext_receive_inject_cloned_bl(
+    _In_ const NET_BUFFER_LIST* cloned_nbl, _In_ const FWPS_INCOMING_VALUES* incoming_fixed_values)
+{
+    uint32_t interface_index =
+        incoming_fixed_values->incomingValue[FWPS_FIELD_INBOUND_MAC_FRAME_NATIVE_INTERFACE_INDEX].value.uint32;
+    uint32_t ndis_port =
+        incoming_fixed_values->incomingValue[FWPS_FIELD_INBOUND_MAC_FRAME_NATIVE_NDIS_PORT].value.uint32;
+
+    return FwpsInjectMacReceiveAsync(
+        _net_ebpf_ext_l2_injection_handle,
+        NULL,
+        0,
+        FWPS_LAYER_INBOUND_MAC_FRAME_NATIVE,
+        interface_index,
+        ndis_port,
+        (NET_BUFFER_LIST*)cloned_nbl,
+        _net_ebpf_ext_l2_receive_inject_complete,
+        NULL);
+}
+
+static void
+_net_ebpf_ext_l2_inject_send_complete(_In_ const void* context, _Inout_ NET_BUFFER_LIST* nbl, BOOLEAN dispatch_level)
+{
+    if ((uintptr_t)context == 0x0)
```

Comment:
why not
```suggestion
    if (context == NULL)
```
?

Reply:
The caller passes a boolean as injection context by force typecasting it to void*. To preserve that semantics, the callback is casting it back to a boolean and checking for `FALSE` instead.